S3UTILS-12: add compareBuckets script

alexanderchan-scality opened the pull request.

alexanderchan-scality committed:
S3UTILS-12: add compareBuckets script

bert-e commented:
# Hello alexanderchan-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/s3utils/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-12 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.9.2`


Please check the `Fix Version/s` of S3UTILS-12, or the target
branch of this pull request.

jonathan-gramain commented on compareBuckets/utils.js:
```diff
@@ -0,0 +1,106 @@
+}
+
+function listBucketMasterKeys(params, cb) {
+    const { bucket, marker, hostPort, maxKeys, workers } = params;
```
Now that I think about it, the listing type can be set with `listingType: 'DelimiterMaster'` in the query params, so that version keys (those with `\0` in them) are not returned. It could make the listing more efficient and remove the need to filter out versions.

jonathan-gramain commented on compareBuckets/utils.js:
```diff
@@ -0,0 +1,106 @@
+        const resp = JSON.parse(res.body);
+        const { Contents, IsTruncated } = resp;
+
+        return async.filterLimit(Contents, workers, (item, itemDone) => {
```
This logic could use the synchronous `Contents.filter(item => ...)`

alexanderchan-scality replied:
adding the `listingType` params to remove this

jonathan-gramain commented on tests/unit/compareBuckets/compareBuckets.js:
```diff
@@ -0,0 +1,411 @@
+        });
+    });
+
+    it('should succesufully perform compare (single listing call)', done => {
```
```suggestion
    it('should successfully perform compare (single listing call)', done => {
```

jonathan-gramain commented on compareBuckets/utils.js:
```diff
@@ -0,0 +1,106 @@
+    return `${baseURL}${queryParams.length > 0 ? '?' : ''}${queryParams.join('&')}`;
+}
+
+function listBucketMasterKeys(params, cb) {
```
We need to wrap this function with a retry loop, otherwise the script will stop at the first error, which is likely to happen on a large bucket

alexanderchan-scality replied:
adding `async.retry` around the `httpRequest`

jonathan-gramain commented on compareBuckets/index.js:
```diff
@@ -0,0 +1,146 @@
+    DST_BUCKET: bucket to be scanned
+
+Optional environment variables:
+    SRC_KEY_MARKER: key to continue listing from on the source bucket
```
I think the user should only pass a single option here to resume from a point, up to the script to set its state and resume the listing from the same point on both bucketds.

jonathan-gramain reviewed (changes requested):
Only the missing retries look critical to me, the rest should be good. I can add the retries and I may also do some other changes.

jonathan-gramain committed:
S3UTILS-12 fixes

- add retries to listing iterations

- set listing type to DelimiterMaster

- Pass a single KEY_MARKER environment variable

- logProgress() only logs a single keyMarker, the lower of the source
  and destination

- replace setTimeout by process.nextTick

- pass missing 'marker' to bucketd params

- remove unused 'workers' param

jonathan-gramain commented on compareBuckets/compareBuckets.js:
```diff
@@ -0,0 +1,140 @@
+
+                srcContents = srcContents.slice(srcIdx);
+                dstContents = dstContents.slice(dstIdx);
+                return setTimeout(() => done(null), 500);
```
Is there a reason for the `setTimeout`? I think it could make things painfully slow on a powerful platform.

alexanderchan-scality replied:
this was a precaution against making rapid requests. I can remove it if this presents itself as a bottleneck

jonathan-gramain commented on compareBuckets/compareBuckets.js:
```diff
@@ -0,0 +1,140 @@
+                            srcContents = contents;
+                            srcDone = !isTruncated;
+                            srcMarker = marker;
+                            statusObj.srcKeyMarker = srcMarker;
```
Here `bucketdSrcParams.marker` should be updated, fixed it to avoid looping over the same range

alexanderchan-scality replied:
thanks! didn't catch this with the mock, and the issue popped when testing against a real backend

alexanderchan-scality commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-12 contains:


* `1.9.2`

* `1.9.3`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.9.4`


Please check the `Fix Version/s` of S3UTILS-12, or the target
branch of this pull request.




*The following options are set:* **approve**

alexanderchan-scality commented:
/help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 






*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **approve**

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-12 contains:


* `1.9.4`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.9.2`


Please check the `Fix Version/s` of S3UTILS-12, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **approve**

alexanderchan-scality committed:
S3UTILS-12: add improvements to reports + remove duplication in tests

alexanderchan-scality commented:
/wait

jonathan-gramain reviewed (approved):
LGTM

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-12 contains:


* `1.9.4`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.10.0`


Please check the `Fix Version/s` of S3UTILS-12, or the target
branch of this pull request.




*The following options are set:* **approve**

alexanderchan-scality commented:
/help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 






*The following options are set:* **approve**

bert-e commented:
# Queue out of order



The changeset has received all authorizations to enter the merge queue,
however the merge queue is currently in an incoherent state. In order
to protect this pull-request, I have not added the changeset to the
queue.

Please contact a member of release engineering.





*The following options are set:* **approve**

alexanderchan-scality commented:
/help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 






*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.10`



The following branches will **NOT** be impacted:

* `development/1.8`
* `development/1.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.10`



The following branches have **NOT** changed:

* `development/1.8`
* `development/1.9`



Please check the status of the associated issue S3UTILS-12.

Goodbye alexanderchan-scality.
